build-tools/packages/build-cli/test/commands/typetests.test.ts

```diff
@@ -2,11 +2,330 @@
  * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
  * Licensed under the MIT License.
  */
-import { assert } from "chai";
+import { assert, expect } from "chai";
+import clonedeep from "lodash.clonedeep";
 
-import { previousVersion } from "../../src/commands/typetests";
+import {
+	TypeTestConfigActions,
+	VersionOptions,
+	previousVersion,
+	updateTypeTestConfiguration,
+} from "../../src/commands/typetests";
+import { PackageJson } from "@fluidframework/build-tools";
+
+/**
+ * A minimal test package.json. It defines only the required fields according to the type definition.
+ */
+const packageMinimal: PackageJson = {
+	name: "test-package",
+	version: "2.0.0-internal.6.0.0",
+	scripts: {},
+};
+
+/**
+ * A test package.json with an enabled typeValidation node.
+ */
+const packageWithTypeValidation: PackageJson = {
+	name: "test-package",
+	version: "2.0.0-internal.6.0.0",
+	scripts: {},
+	devDependencies: {
+		"test-package-previous": "npm:test-package@2.0.0-internal.4.0.0",
+	},
+	typeValidation: {
+		broken: {
+			"broken-package": {
+				backCompat: false,
+				forwardCompat: false,
+			},
+		},
+		disabled: false,
+	},
+};
+
+/**
+ * A test package.json with an disabled typeValidation node.
+ */
+const packageWithTypeValidationDisabled: PackageJson = {
+	name: "test-package",
+	version: "2.0.0-internal.6.0.0",
+	scripts: {},
+	typeValidation: {
+		broken: {
+			"broken-package": {
+				backCompat: false,
+				forwardCompat: false,
+			},
+		},
+		disabled: true,
+	},
+};
+
+const optionsMatrix: Map<string, TypeTestConfigActions> = new Map([
```

Comment:
You are taking a testing strategy here which assumes TypeTestConfigActions.version and TypeTestConfigActions.resetBroken need to be tested together.

If we look at the implementation (thus doing https://en.wikipedia.org/wiki/White-box_testing ), its clear that these two aspects are actually orthogonal and can be tested independently. This does two important things:
1. Allows us to have far fewer tests since we don't need to test combinations.
2. Allows our tests to be simpler since each one only needs to be concerned about one of these two aspects.

If we really wanted to, we could refactor the method we are testing into two separate methods, one for each aspect, but I don't think that is worth it (it would help prevent the tests from becoming out of date if the code was refactored to these two aspect were coupled, but such a change is unlikely, and would require tests updates anyway, and doing such a split just makes more methods/APIs to worry about).

Thus rather than resorting to table driven property testing here, I'd recommend a set of tests for setting the version, and a test for reset broken.

Reply:
Also this setup seems a bit odd to me. Normally you either:
1. define a list or matix of test cases like this, and loop over them generating one test for each which asserts that some properties hold
2. write a test for the different cases, each setting up its inputs and checking it specific requirments hold.

From what I can tell here you have combined these such that the data is seperated from the logic which processes it (case 1), but have seperate logic for each case. This defeats the usecase for case 1 (combining all the logic into one) and makes the tests hard to follow because the logic and case are seperated. This also makes it harder to specialize each test case for the specific things it needs to check. It also makes it easy to have bugs in the testing where cases here are accidently not used below.

This is such a supprising approach that when seeing this list of test cases, I assumed below it would be a loop over them asserting particularl properties hold.

Reply:
> If we really wanted to, we could refactor the method we are testing into two separate methods, one for each aspect,
> but I don't think that is worth it (it would help prevent the tests from becoming out of date if the code was
> refactored to these two aspect were coupled

I know it's not strictly required, but I'm going to do it now in the spirit of making these tests and
accompanying code a "good example".

> This is such a supprising approach that when seeing this list of test cases, I assumed below it would be a loop over
> them asserting particularl properties hold.

I did loop originally, but ran into problems because some combinations needed different assertions. Splitting the
testing into independent pieces will address both of these problems and should make everything more readable.